core/src/main/java/org/elasticsearch/search/fetch/subphase/DocValueFieldsFetchSubPhase.java

```diff
@@ -48,26 +53,42 @@ public void hitExecute(SearchContext context, HitContext hitContext) throws IOEx
                 context.docValueFieldsContext().fields().add(name);
             }
         }
+
         if (context.docValueFieldsContext() == null) {
             return;
         }
+
+        hits = hits.clone(); // don't modify the incoming hits
+        Arrays.sort(hits, (a, b) -> Integer.compare(a.docId(), b.docId()));
+
         for (String field : context.docValueFieldsContext().fields()) {
-            if (hitContext.hit().fieldsOrNull() == null) {
-                hitContext.hit().fields(new HashMap<>(2));
-            }
-            DocumentField hitField = hitContext.hit().getFields().get(field);
-            if (hitField == null) {
-                hitField = new DocumentField(field, new ArrayList<>(2));
-                hitContext.hit().getFields().put(field, hitField);
-            }
             MappedFieldType fieldType = context.mapperService().fullName(field);
             if (fieldType != null) {
-                /* Because this is called once per document we end up creating a new ScriptDocValues for every document which is important
-                 * because the values inside ScriptDocValues might be reused for different documents (Dates do this). */
-                AtomicFieldData data = context.fieldData().getForField(fieldType).load(hitContext.readerContext());
-                ScriptDocValues<?> values = data.getScriptValues();
-                values.setNextDocId(hitContext.docId());
-                hitField.getValues().addAll(values);
+                int currentReaderIndex = -1;
+                LeafReaderContext subReaderContext = null;
+                AtomicFieldData data = null;
+                ScriptDocValues<?> values = null;
+                for (SearchHit hit : hits) {
+                    int readerIndex = ReaderUtil.subIndex(hit.docId(), context.searcher().getIndexReader().leaves());
+                    // if the reader index has changed we need to get a new doc values reader instance
+                    if (readerIndex != currentReaderIndex) {
```

Comment:
you could do `if (subReaderContext == null || hit.docId() >= subReaderContext.docBase + subReaderContext.reader().maxDoc())` to avoid doing the `ReaderUtil.subIndex` binary search for every doc